Update, including new deploy-local.py changes (#239)
bobanetwork/boba: .gitbook/contents/hc/4_deploy-smart-contract.md

# Deploy the Smart Contracts

A [`deploy-local.py`](https://github.com/bobanetwork/rundler-hc/blob/boba-develop/hybrid-compute/deploy-local.py)
script is provided to deploy and configure the necessary contracts on a local stack,
launched from `make devnet-up` in a Boba/Optimism repository.

Usage example:

```bash
cd rundler-hc/hybrid-compute
python ./deploy-local.py --boba-path /home/user/boba  # Supply the path where you checked out the repo
```

This script uses the Foundry toolkit to deploy the base contracts and a set
of examples, via Forge scripts located in `rundler-hc/crates/types/contracts/hc_scripts`.

> The Forge scripts are written in Solidity, with toolkit extensions to support scripting (for example
> reading system environment variables and using them within Solidity functions). Code which executes
> within a block:

Reviewer: ```suggestion
The Forge scripts are written in Solidity, with toolkit extensions to support scripting, such as
reading system environment variables and using them within Solidity functions. Code which executes
within a block (like the following) is translated into a transaction which is first simulated locally and then executed on-chain:
```

```solidity
vm.startBroadcast(deployerPrivateKey);

// ... do stuff

vm.stopBroadcast();
```

> is translated into a transaction which is first simulated locally and then executed on-chain. The
> addresses of newly deployed contracts are returned as an array of values:

Reviewer: ```suggestion
The addresses of newly deployed contracts are then returned as an array of values:
```

l1.middleware_onion.inject(geth_poa_middleware, layer=0)

w3 = Web3(Web3.HTTPProvider(env_vars['NODE_HTTP']))
assert (w3.is_connected)
```

Next, we access the contracts using a `loadContract()` function, such as
for our `HybridAccount` contract:

```python
HA = loadContract(w3, "HybridAccount", path_prefix+"samples/HybridAccount.sol")
```

Here `path_prefix` indicates where the contract's source code is located. The current deployer script
will compile the Solidity code itself to determine the contract ABI, rather than requiring a JSON file
from a previous deployment.

After loading the contract, we can interact with it like this:

```python
def permitCaller(acct, caller):
    if not acct.functions.PermittedCallers(caller).call():
        print("Permit caller {} on {}".format(caller, acct.address))
        tx = acct.functions.PermitCaller(caller, True).build_transaction({
            'nonce': w3.eth.get_transaction_count(env_vars['OC_OWNER']),
            'from': env_vars['OC_OWNER'],
            'gas': 210000,
            'chainId': HC_CHAIN,
        })
        signAndSubmit(w3, tx, env_vars['OC_PRIVKEY'])

# ...

for a in example_addrs:
  permitCaller(HA, a)

```

> Here we are performing a sequence of transactions to add individual example contract addresses to the
> list of contracts which are permitted to use our HybridAccount. There is a check so that this operation
> is only performed for contracts which have not yet been registered.

Reviewer: ```suggestion
Here we perform a sequence of transactions to add individual example contract addresses to the
list of contracts which are permitted to use our `HybridAccount`. There is a check so that this operation
is only performed for contracts which have not yet been registered.
```

    server.register_function(offchain_getprice, selector("getprice(string)"))
```

> You can see how these methods are invoked from the `TestCounter.sol` contract, for example "ramble" is 
> used in a simple word-guessing game:

Reviewer: ```suggestion
You can see how these methods are invoked from the `TestCounter.sol` contract. For example, "ramble" is 
used in a simple word-guessing game:
```
